### DISCUSSION TITLE
Network visualization using VisPy

### DISCUSSION THREAD

**Discussion by ivonindima** [PMEAL/OpenPNM#2414] [Tips and Tricks]
Dear Dr. @jgostick 
I've seen a few questions from the community about using Paraview to visualize networks. Paraview is a great tool, but it takes a few steps to get a good rendering.
I want to share a simple script that allows you to visualize network model using VisPy. The method uses a pseudo-3D approach for drawing spheres, so it is well suited for quick visualization of networks. Performance is especially good when rendering very large networks (one million pores or more). The algorithm is easily extended to visualize transport processes (for example, based on OpenPNM percolation algorithms)
Hope this will be helpful.
Thank you for creating amazing PoreSpy and OpenPNM packages!

Example:
![screenshot](https://user-images.githubusercontent.com/10146235/167010475-eb2ea2f3-0a07-4b9d-8d68-ff11b380b737.png)

```python
import numpy as np
import sys
import vispy
import openpnm as op
from vispy import app, scene
from PyQt5 import QtWidgets
from matplotlib import cm


def create_pore_colors_from_array(a, cmap = 'viridis'):
    colormap = cm.get_cmap(cmap)
    return colormap(a/a.max())

def create_throat_colors_from_array(a, cmap = 'viridis'):
    colormap = cm.get_cmap(cmap)
    return np.repeat(colormap(a/a.max()),2,axis=0)


def create_visual_pnm(pn):
    canvas = scene.SceneCanvas(keys='interactive', show=True)
    view = canvas.central_widget.add_view()
    view.camera = 'turntable'
    view.camera.fov = 30
    view.camera.distance = 3*np.max(pn['pore.coords'])

    pore_color = create_pore_colors_from_array(pn['pore.diameter'], cmap = 'viridis')
    throat_color = create_throat_colors_from_array(pn['throat.diameter'], cmap = 'viridis')
    #plot spheres
    vis = scene.visuals.Markers(
        pos=pn['pore.coords'],
        size=pn['pore.diameter'],
        antialias=1,
        face_color = pore_color,
        edge_width=0,
        scaling=True,
        spherical=True,
    )
    vis.parent = view.scene
    #plot axis
    vispy.scene.visuals.XYZAxis(parent=view.scene)
    #set camera center
    view.camera.center = np.array((pn['pore.coords'][:,0].max()/2, pn['pore.coords'][:,1].max()/2, pn['pore.coords'][:,2].max()/2))
    #data preparation 
    lines = np.zeros((len(pn['throat.conns']), 2,3))
    for i in range(len(pn['throat.conns'])):
        pair = pn['throat.conns'][i]
        line = np.array([[pn['pore.coords'][pair[0]], pn['pore.coords'][pair[1]]]])
        lines[i,:,:] = line
    #plot throats
    vis2 = scene.visuals.Line(lines, width = 2, color = throat_color, connect='segments', antialias=True,)
    vis2.parent = view.scene
    

if __name__ == '__main__':
    import time
    app = QtWidgets.QApplication(sys.argv)
    pn = op.network.Cubic(shape=[30, 30, 30], spacing=[1, 1, 1], connectivity=6)
    geom = op.geometry.GenericGeometry(network=pn, pores=pn.Ps, throats=pn.Ts)
    geom['pore.diameter'] = np.random.rand(pn.Np)  
    print(f'Total number of pores: {pn.Np}')
    print(f'Total number of throats: {pn.Nt}')
    P12 = pn['throat.conns']          
    D12 = geom['pore.diameter'][P12]  
    Dt = np.amin(D12, axis=1)
    geom['throat.diameter'] = Dt
    h = pn.check_network_health()
    #op.topotools.trim(network=pn, pores=h['trim_pores'])

    create_visual_pnm(pn)

    app.exec_() 
```

**Comment by jgostick** [comment]
Wow!  Thanks @ivonindima, I will be trying this out asap.  Thanks very much for sharing.

**Reply by ma-sadeghi** [reply]
@ivonindima That's really cool! Would you be interested in adding the script you shared to OpenPNM under the `contrib` folder? If you could just create a Pull Request and add the script as is, I can do clean it up and fix the `__init__` file, etc. Would greatly appreciate it!

**Reply by jgostick** [reply]
This worked for me on the first try!  I was expecting that the PyQt3 install would be a problem but it was smooth.  Once it's in ``contrib`` we can then clean it up and move it to the main package.  We're using ``contrib`` as a sort of staging area to accept new contributions that may not be 100% finalized yet.  

Note: the PyQt3 library has a GPL license so we may not be able to include it as a dependency.  I wonder if any of our other dependencies have that?  I bet there is a github action that combs for license conflicts.

**Reply by ivonindima** [reply]
Thanks for the answer. I just sent a request.
PyQt5 can be easily replaced by a PySide6 that provided under the LBPL license. Other possible backends for a VisPy are listed [here](https://vispy.org/installation.html).
